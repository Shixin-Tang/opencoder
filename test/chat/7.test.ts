
import { buildComponentTree } from "../utils/debugger.js"
import { queryComponentTree } from "../utils/query.js"
import { waitNextRender } from "../utils/render.js"
import { createMockModel, setupTestEnvironment } from "./util.js"

test("long streaming response renders incrementally and completely", async () => {
  // 1. Setup model with many chunks
  const streamChunks: any[] = []
  const totalChunks = 30
  let fullMessage = ""
  for (let i = 0; i < totalChunks; i++) {
    const text = `Chunk ${i + 1}. `
    streamChunks.push({ type: "text-delta", textDelta: text })
    fullMessage += text
  }
  streamChunks.push({
    type: "finish",
    finishReason: "stop",
    usage: { promptTokens: 5, completionTokens: totalChunks },
  })

  const mockModel = createMockModel(streamChunks)

  const { instance, stdin, stdout, fiber } = await setupTestEnvironment({ model: mockModel })
  assert(stdin)

  // 2. Simulate user input
  stdin.emit("input", "Tell me a long story")
  await waitNextRender()
  stdin.emit("input", "\r")
  await waitNextRender()

  // 3. Observe intermediate states (optional but good for debugging)
  let intermediateOutput = ""
  await vi.waitFor(
    () => {
      intermediateOutput = stdout.get()
      expect(intermediateOutput).toContain("Chunk 1.")
      expect(intermediateOutput).toContain("Chunk 5.")
    },
    { timeout: 500 },
  )

  // 4. Wait for the stream to finish completely
  await vi.waitFor(
    () => {
      const tree = buildComponentTree(fiber!.current.child)
      if (queryComponentTree(tree, "Spinner")) {
        throw new Error("Spinner still rendering")
      }
      const finalOutput = stdout.get()
      expect(finalOutput).toContain("Chunk 1.")
      expect(finalOutput).toContain(`Chunk ${Math.floor(totalChunks / 2)}.`)
    },
    { timeout: 2000 },
  )

  // 5. Verify final output snapshot
  expect(stdout.get()).toMatchSnapshot("long stream - final state")
})
