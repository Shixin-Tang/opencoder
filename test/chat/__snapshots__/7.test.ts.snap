// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`long streaming response renders incrementally and completely > long stream - final state 1`] = `
"> Tell me a long story
▼
 ╭─────────────────────────────────────────────────────────────────────────────────────────────────╮
 │Chunk 1. Chunk 2. Chunk 3. Chunk 4. Chunk 5. Chunk 6. Chunk 7. Chunk 8. Chunk 9. Chunk 10. Chunk │
 │11. Chunk 12. Chunk 13. Chunk 14. Chunk 15. Chunk 16. Chunk 17. Chunk 18. Chunk 19. Chunk 20.    │
 │Chunk 21. Chunk 22. Chunk 23. Chunk 24. Chunk 25. Chunk 26. Chunk 27. Chunk 28. Chunk 29. Chunk  │
 │30.                                                                                              │
 ╰─────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ >                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                    alt + ⏎ for newline ·35 tokens"
`;
